textgrad/engine/ollama.py

```diff
@@ -0,0 +1,100 @@
+try:
+    from ollama import Client
+    # from openai import OpenAI
+except ImportError:
+    raise ImportError("If you'd like to use OpenAI models, please install the openai package by running `pip install openai`, and add 'OPENAI_API_KEY' to your environment variables.")
+
+import os
+import platformdirs
+from tenacity import (
+    retry,
+    stop_after_attempt,
+    wait_random_exponential,
+)
+import json
+from .base import EngineLM, CachedEngine
+from openai._types import NotGiven
+
+class ChatOllama(EngineLM, CachedEngine):
+    DEFAULT_SYSTEM_PROMPT = "You are a helpful, creative, and smart assistant."
+
+    def __init__(
+        self,
+        model_string="qwen2:0.5b",
+        system_prompt=DEFAULT_SYSTEM_PROMPT,
+        **kwargs):
+        """
+        :param model_string:
+        :param system_prompt:
+        """
+        root = platformdirs.user_cache_dir("textgrad")
+        cache_path = os.path.join(root, f"cache_ollama_{model_string}.db")
+
+        super().__init__(cache_path=cache_path)
+
+        self.system_prompt = system_prompt
+        # if os.getenv("OPENAI_API_KEY") is None:
+        #     raise ValueError("Please set the OPENAI_API_KEY environment variable if you'd like to use OpenAI models.")
+        
+        self.client = Client(
+            # api_key="ollama",
+            host="http://localhost:11434",
+        )
+        self.model_string = model_string
+
+
+
+    def generate(
+        self, prompt, system_prompt=None, temperature=0, max_tokens=2000, top_p=0.99
+    ):
+
+        sys_prompt_arg = system_prompt if system_prompt else self.system_prompt
+
+        cache_or_none = self._check_cache(sys_prompt_arg + prompt)
+        if cache_or_none is not None:
+            return cache_or_none
+
+        response = self.client.chat(
+            model=self.model_string,
+            messages=[
+                {"role": "system", "content": sys_prompt_arg},
+                {"role": "user", "content": prompt},
+            ],
+            # frequency_penalty=0,
```

Comment:
can just remove these. although im not too sure w/ ollama -- don't we have temperature / etc. modifiable at inference time?